doc/posixregex.1.in
.\"
.\" DMS Tools and Utilities
.\" Copyright (C) 2010 David M. Syzdek <david@syzdek.net>
.\"
.\" @SYZDEK_COPYRIGHT_HEADER_START@
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.2
.\" or any later version published by the Free Software Foundation;
.\" with no Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.
.\" A copy of the license is included in the section entitled "GNU
.\" Free Documentation License".
.\"
.\" @SYZDEK_COPYRIGHT_HEADER_END@
.\"
.\" doc/posixregex.1.in - man page for posixregex
.\"
.TH "POSIXREGEX" "1" "@RELEASE_MONTH@" "@PACKAGE_TARNAME@ @PACKAGE_VERSION@" "User Commands"
.SH NAME
posixregex \- displays the result of two values after performing bitwise operations
.SH SYNOPSIS
\fBposixregex\fR [-r \fIregex\fR] \fIstring1 string2 ... stringN\fR
.sp
\fBposixregex\fR [\fB--help\fR | \fB-h\fR] [\fB--posixregex\fR | \fB-p\fR] [\fB--version\fR | \fB-V\fR]

.SH DESCRIPTION
The \fBposixregex\fR utility tests a regular expression against the specified
strings and displays the matched segments of the strings.

.SH OPTIONS
.TP
\fB\-h\fR, \fB--help\fR
Displays usage information and exits.
.TP
\fB\-p\fR, \fB--posixregex\fR
Print regular expression special patterns.
.TP
\fB\-q\fR, \fB--quiet\fR, \fB--silent\fR
displays verbose messages.
.TP
\fB\-r\fR
regular expression to use for testing strings
.TP
\fB\-V\fR, \fB--version\fR
displays version number and exits.
.TP
\fB\-v\fR, \fB--verbose\fR
do not print messages.

.SH "ENVIRONMENT VARIABLES"
.TP
\fBPOSIXREGEX\fR
default regular expression to use for testing strings

.SH "DIAGNOSTICS"
Normally, exit status is 0 if the strings match the regular expression and
1 otherwise.

.SH "REPORTING BUGS"
Report bugs to David M. Syzdek <@PACKAGE_BUGREPORT@>.

.SH "SEE ALSO"
.BR bindump (1),
.BR codetagger (1),
.BR numconvert (1)
.\" end of man page


ChangeLog

DMS Tools and Utilities
Copyright (C) 2008, 2009, 2010 David M. Syzdek <david@syzdek.net>

1.0	2010/01/18 (syzdek)
	- Updating bindump to binarily diff files. (syzdek)
	- Adding ANSI color formatting to bindump output. (syzdek)
	- Starting libdms.la for common functions. (syzdek)
	- Adding option in bindump to display bits in little endian (syzdek)
	- Adding option in numconvert to display bits and bytes in little
	  endian (syzdek)
	- Updating bitops to use the same command line flag as numconvert
	  for hex output. (syzdek)
	- Creating initial man pages for bindump, binops, numconvert, and
	  posixregex. (syzdek)
	- Fixing compiler misc warnings. (syzdek)

0.9	2010/01/06 (syzdek)
	- Adding bindump utility (syzdek)

0.8	2009/12/31 (syzdek)
	- Adding typedef-info utility (syzdek)
	- Adding bitcopy utility (syzdek)

0.7	2009/09/23 (syzdek)
	- Adding support for binary input values in numconvert. (syzdek)
	- Adding ability to specify a specific notation to use for parsing
	  input values in numconvert. (syzdek)
	- Updating the Makefile-xcode wrapper to allow custom targets to be
	  executed from within Xcode. (syzdek)
	- Fixing miscellaneous porting errors to WIN32 platforms. (syzdek)

0.0.6	2009/08/12 (syzdek)
	- Adding initial support for gettext. (syzdek)
	- Adding experimental package manager for slackware linux. (syzdek)
	- Adding number system conversion utility numconvert. (syzdek)

0.0.5	2009/02/09 (syzdek)
	- re-branded package as dmstools. (syzdek)
	- migrated from CVS to Git. (syzdek)
	- derived package version from the git repository. (syzdek)
	- adding support for GNU gettext support (syzdek)
	- standardized the name space for codetagger (syzdek)
	- added ability for codetagger to recurse directories (syzdek)
	- updated codetagger to only write data to disk if expanding the
	  tags would result in a change to the file. (syzdek)
	- updated codetagger to use a margin stripped for trailing white
	  space when inserting empty lines from the tag definition (syzdek)

0.0.4	2008/06/29
	- added the Xcode wrapper Makefile-xcode to the distribution (syzdek)
	- replaced calls to mkstemps() with mkstemp() due to the former not
	  being available on Linux systems with GNU libc. (syzdek)
	- added [:blank:] as valid characters to the tag name (syzdek)
	- the end tag now retains the same character case after the
	  the expansion as before. (syzdek)

0.0.3	2008/06/22
	- updated codetagger to show usage information if no parameters were
	  given. [requested by Royce Williams] (syzdek)
	- updated codetagger to warn if tag definitions were not found in
	  the tag file. [requested by Royce Williams] (syzdek)
	- updated codetagger to warn if no tags were found in the processed
	  file. (syzdek)
	- updated codetagger to not replace original file if no tags were
	  found. [requested by Royce Williams] (syzdek)
	- re-numbered the verbose levels (syzdek)
	- added warnings for unknown tags. (syzdek)
	- expanded man page for codetagger (syzdek)

0.0.2	2008/06/22
	- finished documenting the remaining command line options in the man
	  page. (syzdek)
	- added ability code codetagger to set file ownerships of generated
	  files if the utility is run as root. (syzdek)
	- added comment syntax for doxygen parsing. (syzdek)
	- reviewed and improved source code comments. (syzdek)
	- removed dependency of common.h so that utilities can be compiled
	  outside of the package environment. (syzdek)
	- added Xcode over lay for project. (syzdek)
	- added examples for codetagger (syzdek)
	- added ability to specify different strings/characters for the
	  left and right tag markers in codetagger. (syzdek)
	- make misc code clean-ups. (syzdek)
	- included source for vaportar, an utility which creates random
	  tar files that contain multiple files with the same file
	  name.  I know, useless. (syzdek)
	- added utility that test POSIX regular expressions on strings
	  passed as arguments. (syzdek)

0.0.1   2008/06/16
        - initial release of package (syzdek)



src/posixregex.c
/*
 *  DMS Tools and Utilities
 *  Copyright (c) 2008, 2009 David M. Syzdek <david@syzdek.net>.
 *
 *  @SYZDEK_LICENSE_HEADER_START@
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License Version 2 as
 *  published by the Free Software Foundation.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 *
 *  @SYZDEK_LICENSE_HEADER_END@
 */
/**
 *  @file src/posixregex.c simple POSIX regex testing utility/example
 */
/*
 *  Simple Build:
 *     gcc -W -Wall -O2 -c posixregex.c
 *     gcc -W -Wall -O2 -o posixregex   posixregex.o
 *
 *  GNU Libtool Build:
 *     libtool --mode=compile gcc -W -Wall -g -O2 -c posixregex.c
 *     libtool --mode=link    gcc -W -Wall -g -O2 -o posixregex posixregex.lo
 *
 *  GNU Libtool Install:
 *     libtool --mode=install install -c posixregex /usr/local/bin/posixregex
 *
 *  GNU Libtool Clean:
 *     libtool --mode=clean rm -f posixregex.lo posixregex
 */
#define _SCT_SRC_POSIXREGEX_C 1

/////////////
//         //
//  Notes  //
//         //
/////////////

/*
 *  POSIX        ASCII           Description
 *  [:alnum:]    [A-Za-z0-9]     Alphanumeric characters
 *  [:alpha:]    [A-Za-z]        Alphabetic characters
 *  [:blank:]    [ \t]           Space and tab
 *  [:cntrl:]    [\x00-\x1F\x7F] Control characters
 *  [:digit:]    [0-9]           Digits
 *  [:graph:]    [\x21-\x7E]     Visible characters
 *  [:lower:]    [a-z]           Lowercase letters
 *  [:print:]    [\x20-\x7E]     Visible characters and spaces
 *  [:space:]    [ \t\r\n\v\f]   Whitespace characters
 *  [:upper:]    [A-Z]           Uppercase letters
 *  [:xdigit:]   [A-Fa-f0-9]     Hexadecimal digits
 *  [:punct:]    [-!"#$%&'()*+,./:;<=>?@[\\\]_`{|}~]   Punctuation characters
 */

//   Example Uses:
//      posixregex '^([[:alpha:]]+)://(([[:alnum:]]+)(:([[:alnum:]]+)){0,1}@){0,1}([-.a-z]+)(:([[:digit:]]+))*/([-/[:alnum:]]*)(\?(.*)){0,1}$'  http://jdoe:password@www.foo.org:123/path/to/file?query_string
//      posixregex '\$([[:digit:]]+)\.([[:digit:]]{2,2})' 'Your change is $7.45.'


///////////////
//           //
//  Headers  //
//           //
///////////////

#ifdef HAVE_CONFIG_H
#include <config.h>
#include "common.h"
#endif

#include <stdio.h>
#include <getopt.h>
#include <string.h>
#include <stdlib.h>
#include <regex.h>


///////////////////
//               //
//  i18l Support //
//               //
///////////////////

#ifdef HAVE_GETTEXT
#   include <gettext.h>
#   include <libintl.h>
#   define _(String) gettext (String)
#   define gettext_noop(String) String
#   define N_(String) gettext_noop (String)
#else
#   define _(String) (String)
#   define N_(String) String
#   define textdomain(Domain)
#   define bindtextdomain(Package, Directory)
#endif


///////////////////
//               //
//  Definitions  //
//               //
///////////////////

#ifndef PROGRAM_NAME
#define PROGRAM_NAME "posixregex"
#endif
#ifndef PACKAGE_BUGREPORT
#define PACKAGE_BUGREPORT "david@syzdek.net"
#endif
#ifndef PACKAGE_COPYRIGHT
#define PACKAGE_COPYRIGHT ""
#endif
#ifndef PACKAGE_NAME
#define PACKAGE_NAME ""
#endif
#ifndef PACKAGE_VERSION
#define PACKAGE_VERSION ""
#endif

#undef BUFFER_SIZE
#define BUFFER_SIZE 1024

#undef MAX_MATCHES
#define MAX_MATCHES 128

#ifndef PARAMS
#define PARAMS(protos) protos
#endif


//////////////////
//              //
//  Prototypes  //
//              //
//////////////////

/* main statement */
int main PARAMS((int, char **));

/* displays POSIX regular expressions */
void my_posixregex PARAMS((void));

/* displays usage */
void my_usage PARAMS((void));

/* displays usage */
void my_version PARAMS((void));


/////////////////
//             //
//  Functions  //
//             //
/////////////////

/// main statement
/// @param[in] argc   number of arguments
/// @param[in] argv   array of arguments
int main(int argc, char * argv[])
{
   /* declare local vars */
   int          c;
   int          x;
   int          y;
   int          err;
   int          code;
   int          str_len;
   int          quiet;
   int          verbosity;
   int          opt_index;
   char         arg[BUFFER_SIZE];
   char         msg[BUFFER_SIZE];
   char         str[BUFFER_SIZE];
   regex_t      regex;
   regmatch_t   matches[MAX_MATCHES];
   const char * restr;

   /* getopt options */
   static char   short_opt[] = "hpqr:vV";
   static struct option long_opt[] =
   {
      {"help",          no_argument, 0, 'h'},
      {"posixregex",    no_argument, 0, 'p'},
      {"quiet",         no_argument, 0, 'q'},
      {"silent",        no_argument, 0, 'q'},
      {"verbose",       no_argument, 0, 'v'},
      {"version",       no_argument, 0, 'V'},
      {NULL,            0,           0, 0  }
   };

   /* initialize variables */
   code      = 0;
   restr     = NULL;
   quiet     = 0;
   verbosity = 0;
   opt_index = 0;

   /* loops through arguments */
   while((c = getopt_long(argc, argv, short_opt, long_opt, &opt_index)) != -1)
   {
      switch(c)
      {
         case -1:       /* no more arguments */
         case 0:        /* long options toggles */
            break;
         case 'h':
            my_usage();
            return(0);
         case 'p':
            my_posixregex();
            return(0);
         case 'q':
            quiet = 1;
            verbosity = 0;
            break;
         case 'r':
           restr = optarg;
           break;
         case 'V':
            my_version();
            return(0);
         case 'v':
            quiet = 0;
            verbosity++;
            break;
         case '?':
            fprintf(stderr, _("Try `%s --help' for more information.\n"), PROGRAM_NAME);
            return(1);
         default:
            fprintf(stderr, _("%s: unrecognized option `--%c'\n"
                  "Try `%s --help' for more information.\n"
               ),  PROGRAM_NAME, c, PROGRAM_NAME
            );
            return(1);
      };
   };

   if (!(restr))
      restr = getenv("POSIXREGEX");
   if (!(restr))
   {
      fprintf(stderr, _("%s: missing required argument `-- r'\n"
            "Try `%s --help' for more information.\n"
         ), PROGRAM_NAME, PROGRAM_NAME
      );
      return(1);
   };

   /* compile regular expression for later use */
   if (!(quiet))
      printf("regex: %s\n",restr);
   if ((err = regcomp(&regex, restr, REG_EXTENDED|REG_ICASE)))
   {
      regerror(err, &regex, msg, (size_t)BUFFER_SIZE-1);
      printf("%s\n", msg);
      return(1);
   };

   /*
    *  Loops through command line arguments and tests the compiled expressions
    *  against each argument.
    */
   for (x = optind; x < argc; x++)
   {
      /* copies string into buffer and prints to screen */
      strncpy(arg, argv[x], (size_t)BUFFER_SIZE-1);
      if (!(quiet))
         printf("%3i: %s  ==> ", x-optind+1, arg);

      /* tests the buffer against the regular expression */
      if ((err = regexec(&regex, arg, (size_t)MAX_MATCHES, matches, 0)))
      {
         if (!(quiet))
            printf("not found\n");
         code = 1;
      } else {
         if (!(quiet))
            printf(" found\n");
         str_len = 1;

         /* copies sub matches in buffer string */
         if (verbosity)
         {
            for(y = 0; ((y < MAX_MATCHES) && (matches[y].rm_eo > -1)); y++)
            {
               memset(str, 0, (size_t)BUFFER_SIZE);
               if ((str_len = matches[y].rm_eo - matches[y].rm_so))
               {
                  strncpy(str, &arg[matches[y].rm_so], (size_t)str_len);
                  printf("     submatch %i: %s\n", y, str);
               } else if (matches[y].rm_eo > 0) {
                  printf("   submatch %i:\n", y);
               };
            };
            printf("\n");
         };
      };
   };

   /* free regular expression */
   regfree(&regex);

   /* end function */
   return(code);
}


/// displays POSIX regular expressions
void my_posixregex(void)
{
   // TRANSLATORS: The following strings provide a table of POSIX regular
   // expersion key words, key word values, and a description of the key
   // word.  The table is displayed if the program is passed --posixregex
   // on the command line.
   printf(_("POSIX        ASCII           Description\n"
          "[:alnum:]    [A-Za-z0-9]     Alphanumeric characters\n"
          "[:alpha:]    [A-Za-z]        Alphabetic characters\n"
          "[:blank:]    [ \\t]           Space and tab\n"
          "[:cntrl:]    [\\x00-\\x1F\\x7F] Control characters\n"
          "[:digit:]    [0-9]           Digits\n"
          "[:graph:]    [\\x21-\\x7E]     Visible characters\n"
          "[:lower:]    [a-z]           Lowercase letters\n"
          "[:print:]    [\\x20-\\x7E]     Visible characters and spaces\n"
          "[:space:]    [ \\t\\r\\n\\v\\f]   Whitespace characters\n"
          "[:upper:]    [A-Z]           Uppercase letters\n"
          "[:xdigit:]   [A-Fa-f0-9]     Hexadecimal digits\n"
          "[:punct:]    [-!\"#$%%&'()*+,./:;<=>?@[\\]_`{|}~]   Punctuation characters\n"
          "\n"
          "Example Uses:\n"
          "    posixregex -r '^([[:alpha:]]+)://(([[:alnum:]]+)(:([[:alnum:]]+)){0,1}@){0,1}([-.a-z]+)(:([[:digit:]]+))*/([-/[:alnum:]]*)(\\?(.*)){0,1}$'  http://jdoe:password@www.foo.org:123/path/to/file?query_string\n"
          "    posixregex -r '\\$([[:digit:]]+)\\.([[:digit:]]{2,2})' 'Your change is $7.45.'\n"
      )
   );
   return;
}


/// displays usage
void my_usage(void)
{
// TRANSLATORS: The following strings provide usage for command. These
   // strings are displayed if the program is passed `--help' on the command
   // line. The two strings referenced are: PROGRAM_NAME, and
   // PACKAGE_BUGREPORT
   printf(_("Usage: %s [OPTIONS] string1 string2 ... stringN\n"
         "  -h, --help                print this help and exit\n"
         "  -p, --posixregex          print regular expression patterns\n"
         "  -q, --quiet, --silent     do not print messages\n"
         "  -r regex                  regular expression to use for testing strings\n"
         "  -v, --verbose             print verbose messages\n"
         "  -V, --version             print version number and exit\n"
         "\n"
         "Variables:\n"
         "  POSIXREGEX                default regular expression to use for testing strings\n"
         "\n"
         "Report bugs to <%s>.\n"
      ), PROGRAM_NAME, PACKAGE_BUGREPORT
   );
   return;
}


/// displays version
void my_version(void)
{
   // TRANSLATORS: The following strings provide version and copyright
   // information if the program is passed --version on the command line.
   // The three strings referenced are: PROGRAM_NAME, PACKAGE_NAME,
   // PACKAGE_VERSION.
   printf(_("%s (%s) %s\n"
         "Written by David M. Syzdek.\n"
         "\n"
         "%s\n"
         "This is free software; see the source for copying conditions.  There is NO\n"
         "warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n"
      ), PROGRAM_NAME, PACKAGE_NAME, PACKAGE_VERSION, PACKAGE_COPYRIGHT
   );
   return;
}

/* end of source file */
